src/models/state/wallet/wallet_state.rs

```diff
@@ -225,18 +228,37 @@ impl WalletState {
         Ok(ret)
     }
 
-    pub async fn new_from_wallet_secret(
+    /// Create a `WalletState` object from related data.
+    ///
+    /// Convenience method to extract required data prior to calling the
+    /// canonical constructor, [`Self::new`].
+    pub(crate) async fn new_from_context(
         data_dir: &DataDirectory,
-        wallet_secret: WalletSecret,
+        wallet_file_context: WalletFileContext,
         cli_args: &Args,
     ) -> Self {
-        // Create or connect to wallet block DB
+        let database_is_new = tokio::fs::try_exists(&data_dir.wallet_database_dir_path())
+            .await
+            .unwrap();
+        let configuration = WalletConfiguration::new(data_dir)
+            .absorb_options(cli_args)
+            .with_scan_mode_if_necessary(wallet_file_context.wallet_is_new, database_is_new);
```

Comment:
Would it make sense to move the logic for maybe turning on scan mode in here? That way, the scan mode configuration doesn't need to know about wallet creation and database things.

```rust
// here:
let mut configuration = WalletConfiguration::new(data_dir).absorb_options(cli_args);

// <comment explaining what's going on>
if !wallet_file_context.wallet_is_new && database_is_new {
    configuration.enable_scan_mode();
}

// And in wallet_configuration.rs:
impl WalletConfiguration {
    /// Enables scan mode if it isn't already.
    pub(crate) fn enable_scan_mode(&mut self) {
        if self.scan_mode.is_none() {
            self.scan_mode = Some(ScanModeConfiguration::default());
        }
    }
}
```